### PR TITLE
CSHARP-2512: Change the  `RetryWrites` default value.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**
Spec: https://github.com/mongodb/specifications/commit/ae12fd2921f92608ad953204b0b280c369b8486d
Evergreen: https://evergreen.mongodb.com/version/5cde762a3e8e865856243e1e

### CODE REVIEW

File: tests/MongoDB.Driver.Tests/Specifications/command-monitoring/TestRunner.cs [hunk 1]
```diff
@@ -105,104 +90,169 @@ public void RunTestDefinition(IEnumerable<BsonDocument> data, string databaseNam
         {
             definition = (BsonDocument)DeepCopy(definition); // protect against side effects when the same definition is run twice (async=false/true)
 
-            BsonValue bsonValue;
-            if (definition.TryGetValue("ignore_if_server_version_greater_than", out bsonValue))
+            using (var client = CreateDisposableClient(definition, __capturedEvents))
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/Specifications/command-monitoring/TestRunner.cs]: this spec tests also require explicit setting `retryWrites` to `false`.
The only change here is replacing the global `__client` variable on creating `disposable` client for each test and parsing `clientOptions` section.

File: tests/MongoDB.Driver.Tests/Specifications/command-monitoring/TestRunner.cs [hunk 2]
```diff
@@ -93,6 +92,7 @@ public bool OneTimeSetup()
                     cb.ConfigureServer(ss => ss.With(heartbeatInterval: Timeout.InfiniteTimeSpan));
                 }
             };
+            settings.RetryWrites = false;
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Tests/Specifications/command-monitoring/TestRunner.cs:95]: Otherwise, this test fails with the following error:
```Expected object to be {
	delete  = test,
	deletes = [{
			"q": {
				"_id": {
					"$nothing": 1
				}
			},
			"limit": 1
		}
	],
	ordered = true,
	$db = command - monitoring - tests
}, but found {
	delete  = test,
	ordered = true,
	txnNumber = 28,
	$db = command - monitoring - tests,
	deletes = [{
			"q": {
				"_id": {
					"$nothing": 1
				}
			},
			"limit": 1
		}
	]
}
```

File: tests/MongoDB.Driver.TestHelpers/DriverTestConfiguration.cs [hunk 3]
```diff
@@ -140,6 +140,8 @@ public static MongoClientSettings GetClientSettings()
             {
                 serverSelectionTimeoutString = "30000";
             }
+
+            clientSettings.RetryWrites = false;
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.TestHelpers/DriverTestConfiguration.cs]: Tests that are are not related with `RetryWrites` directly were designed with considering that `RetryWrites` is equal to `false`.

**Comment by DmitryLukyanov** [comment]
NOTE: the spec diff contains more files than are presented in the master for now. For example, we don't have the following files (.json/.yml):

>deleteMany.
>updateMany

 that will be added in the ticket CSHARP-2419

File: src/MongoDB.Driver.Legacy/MongoServerSettings.cs [hunk 4]
```diff
@@ -96,7 +96,7 @@ public MongoServerSettings()
             _readPreference = ReadPreference.Primary;
             _replicaSetName = null;
             _retryReads = true;
-            _retryWrites = false;
+            _retryWrites = true;
```
Comment (jyemin) [inline src/MongoDB.Driver.Legacy/MongoServerSettings.cs:99]: Also update public API documentation to indicate the new default value.

Also, @rstam, is there any place in the reference documentation where we add information about upgrading, similar to http://mongodb.github.io/mongo-java-driver/3.10/upgrading/.  I couldn't find anything, but this would be something I'd add.  If not, maybe add to the `what_is_new` page for the 2.9 release?

File: src/MongoDB.Driver/MongoClientSettings.cs [hunk 5]
```diff
@@ -94,7 +93,7 @@ public MongoClientSettings()
             _readPreference = ReadPreference.Primary;
             _replicaSetName = null;
             _retryReads = true;
-            _retryWrites = false;
+            _retryWrites = true;
```
Comment (jyemin) [inline src/MongoDB.Driver/MongoClientSettings.cs:96]: Same as above: update api doc string

**Review by jyemin — CHANGES_REQUESTED** [review]

**Comment by rstam** [comment]
The JSON tests in this pull request do not match the latest tests in the specifications repo.

Can you identify which commit in the specifications repo the tests match?

And if none, how did you come up with these tests? Did you hand modify them?

Perhaps we are implementing these CSHARP tickets in the wrong order?

**Comment by DmitryLukyanov** [comment]
>The JSON tests in this pull request do not match the latest tests in the specifications repo.
yes
>Can you identify which commit in the specifications repo the tests match?

no, I've added only test changes which are related to the current ticket and which don't require the implementation of any other csharp tickets. For example, the lost changes which @rstam means for this ticket requires the following:
1. Implementation of the ticket Csharp-2419 which adds two new test files which also could be updated by this ticket logic.
2. Implementation of the ticket Csharp-2413 which adds some additional lines in the spec tests. These additional lines are skipped in the current PR.
3. Implementation of the ticket SPEC-1175. I don't know which c# ticket was expected to be an implementation to this ticket originally. But this spec ticket has some renaming which is also not presented in this PR.

My original intention was to write in the other related csharp tickets the comments for example since the ticket hasn't been implemented yet, the following lines from spec for 2512 were skipped. So, as soon as the not implemented ticket would be ready, please also update the code for the ticket 2512.

>Did you hand modify them?

yes

>Perhaps we are implementing these CSHARP tickets in the wrong order?

Order changing will help, but for example, I still don't know in the scope of the which ticket we need to implement the SPEC-1175?

However, the "lost" spec changes are small and I can add them to this PR too. The only question I have should I fully implemented the related CSHARP tickets? For example the ticket CSHARP-2413 not only with SPEC file which are related to this ticket. There are 3 spec tickets inside:
1. **SPEC-870** contains a lot of spec tickets inside, but the changes for the current ticket are only in the  _SPEC-1136_.
2. **SPEC-900**, 
3. **SPEC-1142**.

@jyemin could you please take a look at it?

**Comment by rstam** [comment]
I don't think you should add work to this pull request that is unrelated to the default RetryWrites value.

I just needed to know what was going on with the JSON test files. It is OK if they were hand crafted just for the purpose of this ticket.

Eventually we will resync with the latest version from the specification repo and then we will be back in sync.

I just needed to know what was going on.

File: src/MongoDB.Driver.Legacy/MongoServerSettings.cs (hunk 4 above)
Comment (rstam) [reply src/MongoDB.Driver.Legacy/MongoServerSettings.cs:99]: Information about upgrading issues in general (and related to retryWrites in particular) could be added to this page:

http://mongodb.github.io/mongo-csharp-driver/2.8/upgrading/

File: tests/MongoDB.Driver.TestHelpers/DriverTestConfiguration.cs (hunk 3 above)
Comment (rstam) [reply tests/MongoDB.Driver.TestHelpers/DriverTestConfiguration.cs]: This doesn't feel right. This configuration affects ALL tests.

We should let RetryWrites have its default value and fix the small number of tests that might require overridding the new default value.

File: tests/MongoDB.Driver.Tests/Specifications/command-monitoring/TestRunner.cs (hunk 2 above)
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Specifications/command-monitoring/TestRunner.cs:95]: This doesn't feel right. This configuration affects ALL command monitoring tests.

We should let RetryWrites have its default value and fix the small number of tests that might require overridding the new default value. In fact, when an override is necessary for the test to pass shouldn't that override be part of the JSON test itself?

**Review by rstam — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Tests/Specifications/command-monitoring/TestRunner.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/command-monitoring/TestRunner.cs:95]: In this case, we will need to update json test files which were not updated by the spec team. I thought it has been done in this way because these tests were done for `retryableWrites = false`, or at least they should not be checked with `true` value?
Also, this is how it's implemented in Java: https://github.com/jyemin/mongo-java-driver/blob/4bde0d452ad88c701eab3aeef04e0752b086acb3/driver-sync/src/test/functional/com/mongodb/client/CommandMonitoringTest.java#L85

File: tests/MongoDB.Driver.TestHelpers/DriverTestConfiguration.cs (hunk 3 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.TestHelpers/DriverTestConfiguration.cs]: I was doing it with taking into account the idea that all tests were written with the expected `retryWrites` default value as `false`, except for the tests which specify this parameter explicitly.

But, probably here you are right . 
Changed.

File: tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs [hunk 6]
```diff
@@ -1230,15 +1229,19 @@ public void TestFindAndRemoveWithFields()
         [Fact]
         public void TestFindAndRemoveWithMaxTime()
         {
-            if (_primary.Supports(FeatureId.MaxTime))
+            var server = GetServerFromDisposableClient();
```
Comment (DmitryLukyanov) [inline tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs]: I didn't `dispose` this client because in this case, next test call would fail with the error: 

    System.ObjectDisposedException : Cannot access a disposed object.
    Object name: 'MultiServerCluster'

It's related to the fact that a `disposable client` and `LegacyTestConfiguration.Server` use the same MongoClient static dictionary:
https://github.com/mongodb/mongo-csharp-driver/blob/7cd604ab4fee1c3a02fd56e7b3f99215055785f4/src/MongoDB.Driver.Legacy/MongoServer.cs#L64

Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs]: To use `using` we need to remove disposed server via `MongoServer.UnregisterServer(server)`, not sure that it's a better way than the currently implemented

**Comment by DmitryLukyanov** [comment]
Evergreen: https://evergreen.mongodb.com/version/5ce47f93c9ec440a0c2be201

File: tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs [hunk 7]
```diff
@@ -1252,6 +1274,23 @@ private void CheckExpectedResult(ExpectedResult expectedResult, BulkWriteResult<
...
+
+        private MongoServer GetServerFromDisposableClient()
+        {
+            DisposableMongoClient client = DriverTestConfiguration.CreateDisposableClient(
+                settings =>
+                {
+                    settings.RetryWrites = false;
+                });
+            return client.GetServer();
+        }
```
Comment (rstam) [inline tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs]: I don't think we should be using DisposableMongoClient in this context (we should only use one when we intend to Dispose it).

Remove this method, and instead let's add the following method to LegacyTestConfiguration.cs instead:

            public static MongoServer GetServer(
                bool? retryReads = null,
                bool? retryWrites = null)
            {
                var settings = DriverTestConfiguration.GetClientSettings();
                if (retryReads.HasValue)
                {
                    settings.RetryReads = retryReads.Value;
                }
                if (retryWrites.HasValue)
                {
                    settings.RetryWrites = retryWrites.Value;
                }
                var client = new MongoClient(settings);
    #pragma warning disable 618
                return client.GetServer();
    #pragma warning restore 618
            }

Note: you could simplify it a bit by removing retryReads since only retryWrites is needed so far.

File: tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs [hunk 8]
```diff
@@ -202,28 +201,48 @@ public void TestExecuteTwice(bool ordered)
             Assert.Throws<InvalidOperationException>(() => bulk.Execute());
         }
 
+        public static MongoClientSettings GetClientSettings()
```
Comment (rstam) [inline tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs]: Remove this method. Nothing calls it.

File: tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs [hunk 9]
```diff
@@ -202,28 +201,48 @@ public void TestExecuteTwice(bool ordered)
...
+        }
+
         [Theory]
         [InlineData(false, 0)]
         [InlineData(false, 1)]
         [InlineData(true, 0)]
         [InlineData(true, 1)]
         public void TestExecuteWithExplicitWriteConcern(bool ordered, int w)
         {
+            var server = GetServerFromDisposableClient();
```
Comment (rstam) [inline tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs]: Do this instead:

    var server = LegacyTestConfiguration.GetServer(retryWrites: false);

File: tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs [hunk 10]
```diff
@@ -1127,18 +1146,21 @@ public void TestUpsertWithOneMatchingDocument(bool ordered)
         [InlineData(true)]
         public void TestW0DoesNotReportErrors(bool ordered)
         {
+            var server = GetServerFromDisposableClient();
```
Comment (rstam) [inline tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs]: Do this instead:

    var server = LegacyTestConfiguration.GetServer(retryWrites: false);

File: tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs (hunk 6 above)
Comment (rstam) [reply tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs]: Do this instead:

    var server = LegacyTestConfiguration.GetServer(retryWrites: false);

File: tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs [hunk 11]
```diff
@@ -3577,6 +3580,30 @@ private void EnsureCollectionExists(string collectionName)
...
+                .GetDatabase(CoreTestConfiguration.DatabaseNamespace.DatabaseName)
+                .GetCollection(GetType().Name);
+        }
+
+        private MongoServerInstance GetPrimary(MongoServer server)
+        {
+            return server.Instances.First(c => c.IsPrimary);
+        }
+
+        private MongoServer GetServerFromDisposableClient()
```
Comment (rstam) [inline tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs]: Remove this method.

File: tests/MongoDB.Driver.Tests/Specifications/command-monitoring/TestRunner.cs (hunk 2 above)
Comment (rstam) [reply tests/MongoDB.Driver.Tests/Specifications/command-monitoring/TestRunner.cs:95]: OK. Since we are not updating these JSON tests (yet) we can set RetryWrites to false.

**Review by rstam — CHANGES_REQUESTED** [review]

File: tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs (hunk 6 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs]: Done

File: tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs (hunk 11 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Legacy.Tests/MongoCollectionTests.cs]: Done

File: tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs (hunk 8 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs]: I didn't notice that it wasn't called by me anymore. Done

File: tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs (hunk 9 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs]: Done

File: tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs (hunk 10 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs]: Done

File: tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs (hunk 7 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Legacy.Tests/Operations/BulkWriteOperationTests.cs]: Done

File: Docs/reference/content/upgrading.md [hunk 12]
```diff
@@ -11,9 +11,6 @@ title = "Upgrading"
 
 ## Breaking Changes
 
-For most users there should be no breaking changes in version 2.8.0 of the driver.
+There should be no breaking changes in version 2.9.0 of the driver.
```
Comment (DmitryLukyanov) [inline Docs/reference/content/upgrading.md]: As I see, the default values change is not a breaking change (Since it will not be a reason for app fail). 
If some other tickets change it, this line should be updated.

File: Docs/reference/content/upgrading.md [hunk 13]
```diff
@@ -11,9 +11,6 @@ title = "Upgrading"
...
 
-Because we have updated several external dependencies to newer versions, you might encounter compatibility
-issues if your application depends on different versions of those dependencies. This might apply to you
-if you depend on System.Runtime.InteropServices.RuntimeInformation or DnsClient. Note that even if your
-application does not depend on these directly, it is possible that you depend on them indirectly.
+Starting with the 2.9.0 release, retryable writes are enabled by default.
```
Comment (DmitryLukyanov) [inline Docs/reference/content/upgrading.md]: Since it's not a breaking change (I think so for now), should some title be added before this line?
Like probably:
`### Minor Changes`?

File: src/MongoDB.Driver.Legacy/MongoServerSettings.cs (hunk 4 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver.Legacy/MongoServerSettings.cs:99]: Done

File: src/MongoDB.Driver/MongoClientSettings.cs (hunk 5 above)
Comment (DmitryLukyanov) [reply src/MongoDB.Driver/MongoClientSettings.cs:96]: Done

**Comment by DmitryLukyanov** [comment]
Ready for review.

**Comment by DmitryLukyanov** [comment]
Ready for review again.

**Review by rstam — APPROVED** [review]

File: Docs/reference/content/what_is_new.md [hunk 14]
```diff
@@ -8,6 +8,12 @@ title = "What's New"
   pre = "<i class='fa fa-star'></i>"
 +++
 
+## What's New in 2.9.0
+
+Some of the changes in 2.9.0 include:
+
+* Retryable writes are enabled by default
```
Comment (vincentkam) [inline Docs/reference/content/what_is_new.md:15]: Just noting that if a user has custom retry logic and uses the old default of false, then after upgrading to a version that uses the default of true, their app could end up retrying twice. Probably not worth classing as a breaking change though...

**Review by vincentkam — APPROVED** [review]
LGTM! 😺

**Review by jyemin — APPROVED** [review]
LGTM
